### ISSUE TITLE
ask

### ISSUE THREAD

**Issue by CW29999** [FanShiYuan99/Need-help-with-PINN-and-DeepONet#1]
请问这个问题解决了吗？

**Comment by FanShiYuan99** [comment]
还没有解决呢，可以看我刚更新的[Modified_MLP.html](https://github.com/FanShiYuan99/Need-help-with-PINN-and-DeepONet/blob/main/Modified_MLP.html)，改用WangSifan的Modified_MLP_Block能够用50的样本量达到位移1e-4，应力1e-2的MSE。但加pde_loss哪里错了还没有弄清楚。

**Comment by CW29999** [comment]
我有个疑问，def get_data_Loss(self, B_data, T_data, u_data, S_data):
        y_rand = (torch.rand_like(T_data[:, 1:2]) * 2 - 1) * self.H
        x_bc = torch.full_like(y_rand, -self.L)
        tc_data = torch.cat([x_bc, y_rand], dim=1).detach()
        tc_data.requires_grad = True
        tc_data = tc_data.to(device)
        
        u_pre, v_pre, sxx_pre, syy_pre, sxy_pre = self.u_function(B_data, T_data)
        u_bc, v_bc, _, _, _ = self.u_function(B_data, tc_data)
        u_ex = u_data[:,0:1]
        v_ex = u_data[:,1:2]
        sxx_ex = S_data[:,0:1]
        syy_ex = S_data[:,1:2]
        sxy_ex = S_data[:,2:3]
        loss = torch.mean(torch.square(u_pre - u_ex)) + torch.mean(torch.square(v_pre - v_ex)) + \
        torch.mean(torch.square(u_bc)) + torch.mean(torch.square(v_bc)) + \
        torch.mean(torch.square(sxx_pre-sxx_ex)) + torch.mean(torch.square(syy_pre-syy_ex)) + torch.mean(torch.square(sxy_pre-sxy_ex))
        return loss
这里y_rand的含义是什么？不过我并非力学专业，对该方程并不了解，可能与其数据有关，但是比较疑惑这里的rand方法是在处理什么？

**Comment by FanShiYuan99** [comment]
这里是想增加悬臂梁左边界条件增强loss信息，也就是在x=-L，y∈[-H,H]的点，预测出来的位移u、v应该是0，也就是torch.mean(torch.square(u_bc)) + torch.mean(torch.square(v_bc))这个损失，可以参考[Modified_MLP.html]里我对代码整理了一下。

**Comment by CW29999** [comment]
大致了解了，那你在pideeponet中使用的样本数据大小如何呢？我目前在探究麦克斯韦方程的求解问题，所以想请教一下

**Comment by FanShiYuan99** [comment]
我目前尝试了纯数据驱动，在这个情景下的总样本量[50,125,250,500]，取80%做训练集其余做测试，每个样本有200*50的均匀网格，和几种线弹和超弹模型，最好也就位移1e-4\~1e-5，应力1e-2\~1e-3的MSE。不过我看“https://doi.org/10.1007/s00521-024-10132-2”提到：数据集大小、网格精度、网络和优化器超参数等等几个因素都对DeepONet影响不大，全批次比小批次迭代好。所以若使用piDeepONet的话，理论上不需要太多高精度数据。

**Comment by CW29999** [comment]
好的，谢谢，不知能否拿到一些你代码中的数据样本，进行代码测试

**Comment by FanShiYuan99** [comment]
这里我放了250样本量的线弹性本构模型，杨氏模量1000，泊松比0.3的数据，你看看能不能下载。https://www.kaggle.com/datasets/diabolism66/deeponet-elastic-2d-250/data

**Comment by CW29999** [comment]
好的，非常感谢

**Comment by CW29999** [comment]
我在使用autograd计算二阶导数的时候发现会占用巨大的内存，使用大batch时直接OOM，博主有没有碰到过此类问题

**Comment by FanShiYuan99** [comment]
是的，高阶导数很难求所以我的网络设置的直接输出位移和应力，而不是通过位移的导数求得应力，这样计算pde误差时就最多只用求到二阶导数。我的batch设置的是10000个点，你的gpu显存不够的话得用更小批次采用梯度累计或者移到cpu上计算。

**Comment by CW29999** [comment]
你设置的10000个点为什么在枝干网络里的输入为202，还是说你枝干网络里采样的参数只有这么多？

**Comment by FanShiYuan99** [comment]
之前输入一个参数t效果不好，我怀疑一个维度的负载t提供的信息不够，并且想后面可能在网络增加卷积或者其他什么模块，所以[DeepONet_Elastic_2D_250.html]中尝试的是：Bratch采样悬臂梁右端的x方向和y方向的位移代替1个维度的负载，即X=L位置的共101个点的值*2=202。后面发现两种branch输入方式区别不大，[DeepONet_NeoHookean_2D_500.html]中的Branch输入1个维度也就是负载t，这个是刚跑出来的比较不错的纯数据驱动下的结果。

**Comment by FanShiYuan99** [comment]
请问物理驱动可以做吗？我又做了几个纯数据驱动的结果，在：[MultiParams](https://github.com/FanShiYuan99/Need-help-with-PINN-and-DeepONet/tree/main/MultiParams)。我现在需要添加物理驱动或者其他的什么方法结构进去来增强我的精度和计算效率，请问你这边有什么建议么。

**Comment by CW29999** [comment]
我这边也没有正确的结果，数据驱动我之前已经尝试过，但是在加上PDE损失后，效果比你的demo中更差，基本趋向于0解了，在物理数据混合驱动下，我的data_loss基本降不下来，但是我看你之前demo里，二者是在波动的，你可能需要调整二者的权重？这类方法也有很多，但是我并不太知道怎么使用。可以的话，可以加个联系方式交流
